util/system/path.go

```diff
@@ -16,3 +25,180 @@ func DefaultPathEnv(os string) string {
+	}
+	// We're normalizing the path for a unix system, so convert to slash.
+	return filepath.ToSlash(wd), nil
+}
+
+// IsAbs returns a boolean value indicating whether or not the path
+// is absolute. On Linux, this is just a wrapper for filepath.IsAbs().
+// On Windows, we strip away the drive letter (if any), clean the path,
+// and check whether or not the path starts with a filepath.Separator.
+// This function is meant to check if a path is absolute, in the context
+// of a COPY, ADD or WORKDIR, which have their root set in the mount point
+// of the writable layer we are mutating. The filepath.IsAbs() function on
+// Windows will not work in these scenatios, as it will return true for paths
+// that:
+//   - Begin with drive letter (DOS style paths)
+//   - Are volume paths \\?\Volume{UUID}
+//   - Are UNC paths
+func IsAbs(pth, platform string) bool {
+	if platform == "" {
+		platform = runtime.GOOS
+	}
+	cleanedPath, err := CheckSystemDriveAndRemoveDriveLetter(pth, platform)
+	if err != nil {
+		return false
+	}
+	cleanedPath = filepath.ToSlash(cleanedPath)
+	// We stripped any potential drive letter and converted any backslashes to
+	// forward slashes. We can safely use path.IsAbs() for both Windows and Linux.
+	return path.IsAbs(cleanedPath)
+}
+
+// CheckSystemDriveAndRemoveDriveLetter verifies and manipulates a Windows path.
+// For linux, this is a no-op.
+//
+// This is used, for example, when validating a user provided path in docker cp.
+// If a drive letter is supplied, it must be the system drive. The drive letter
+// is always removed. It also converts any backslash to forward slash. The conversion
+// to OS specific separator should happen as late as possible (ie: before passing the
+// value to the function that will actually use it). Paths are parsed and code paths are
+// triggered starting with the client and all the way down to calling into the runtime
+// environment. The client may run on a foreign OS from the one the build will be triggered
+// (Windows clients connecting to Linux or vice versa).
+// Keeping the file separator consistent until the last moment is desirable.
+//
+// We need the Windows path without the drive letter so that it can ultimately be concatenated with
+// a Windows long-path which doesn't support drive-letters. Examples:
+// C:			--> Fail
+// C:somepath   --> somepath // This is a relative path to the CWD set for that drive letter
+// C:\			--> \
+// a			--> a
+// /a			--> \a
+// d:\			--> Fail
+//
+// UNC paths can refer to multiple types of paths. From local filesystem paths,
+// to remote filesystems like SMB or named pipes.
+// There is no sane way to support this without adding a lot of complexity
+// which I am not sure is worth it.
+// \\.\C$\a     --> Fail
+func CheckSystemDriveAndRemoveDriveLetter(path string, platform string) (string, error) {
+	if platform == "" {
+		platform = runtime.GOOS
+	}
+
+	if platform != "windows" {
+		return path, nil
+	}
```

Comment:
I think the previous impl was better with `CheckSystemDriveAndRemoveDriveLetter(path string)` in `path_unix.go` and `path_windows.go` files. Can we keep that?

Reply:
Windows hosts can theoretically build Linux images, as it also has support for LCOW (Linux Containers on Windows). We should also be able to build Windows images on Linux. It's a good idea to use the target platform as an indicator of how to handle paths (as well as other things).

In short, the platform we're running on is not necessarily the platform we're building for.

@tonistiigi expressed a similar concern here:  https://github.com/moby/buildkit/pull/3322#discussion_r1130188645

Reply:
There was a similar [discussion in containerd](https://github.com/containerd/containerd/pull/8362#discussion_r1194080719). Which is why, in that instance, we ended up using the ```image.ImageSpec.OS``` as an indicator of how to handle the volumes when copying existing contents.